src/Compilers/CSharp/Test/EndToEnd/EndToEndTests.cs

```diff
@@ -514,5 +514,125 @@ string makeExpectedOutput()
+            }
+
+            var builder3 = new StringBuilder();
+            // Local functions should be similarly fast as lambdas.
+            if (localFunctions)
+            {
+                /*
+                    M2(x, L0);
+                    static I0 L0(string arg) {
+                        arg = arg + "0";
+                        // ...
+                        arg = arg + "9";
+                        return default;
+                    }
+                    // ...
+                    M2(x, L9);
+                    static I0 L9(string arg) { ... }
+                */
+                for (int i = 0; i < lambdasNumber; i++)
+                {
+                    builder3.AppendLine($"M2(x, L{i});");
+                    builder3.AppendLine($$"""static I0 L{{i}}(string arg) {""");
+                    for (int j = 0; j < statementsNumber; j++)
+                    {
+                        builder3.AppendLine($"""arg = arg + "{j}";""");
+                    }
+                    builder3.AppendLine("return default;");
+                    builder3.AppendLine("}");
+                }
+            }
+            else
+            {
+                /*
+                    M2(x, static I0 (string arg) => {
+                        arg = arg + "0";
+                        // ...
+                        arg = arg + "9";
+                        return default;
+                    });
+                    // ...
+                    M2(x, static I0 (string arg) => {
+                        // ...
+                        return default;
+                    });
+                */
+                for (int i = 0; i < lambdasNumber; i++)
+                {
+                    builder3.AppendLine("M2(x, static I0 (string arg) => {");
+                    for (int j = 0; j < statementsNumber; j++)
+                    {
+                        builder3.AppendLine($"""arg = arg + "{j}";""");
+                    }
+                    builder3.AppendLine("return default;");
+                    builder3.AppendLine("});");
+                }
+            }
+
+            var source = $$"""
+                {{builder1}}
+                class C
+                {
+                    {{builder2}}
+                    void Main()
+                    {
+                        M1(x =>
+                        {
+                            {{builder3}}
+                        });
+                    }
+                }
+                """;
+            RunInThread(() =>
+            {
+                var comp = CreateCompilation(source, options: TestOptions.DebugDll.WithConcurrentBuild(false));
+                var data = new LambdaBindingData();
+                comp.TestOnlyCompilationData = data;
+                comp.VerifyDiagnostics();
+                Assert.Equal(localFunctions ? 20 : 220, data.LambdaBindingCount);
+            }, timeout: TimeSpan.FromSeconds(20));
```

Comment:
This runs ~45 seconds (before the fix) vs ~5 seconds (after the fix) on my machine for lambdas and ~5 seconds both before and after for local functions.

Also without the fix, the number of lambda bindings is 1930 instead of 220.